woocommerce/google-listings-and-ads
# PMax Assets: Preprocessing for API integrations

### Changes proposed in this Pull Request:

This PR implements the preprocessing for 📌 [Integrating with backend APIs](https://github.com/woocommerce/google-listings-and-ads/issues/1787#fe-integrating-apis) in #1787.

In order to avoid the subsequent PR of the API integrations being too large, some are split into preprocessing:

- Fix that the submission of "Skip adding assets" button is blocked when creating and updating campaigns.
- Rename the **asset-specs.js** file to **assetSpecs.js** since most non-component file names were using camel case naming.
- Remove a misplaced test from the **validateAssetGroup.test.js** (https://github.com/woocommerce/google-listings-and-ads/commit/2185ff13acfcf100868d6122ed8e0c3f55c3efb9)
- Create constants for the keys of asset group and related data structures as they will be wildly used later.
- Adjust the type definitions of asset group and related data structures to map to the response of asset groups and suggested assets. (Examples of the different API responses are shown in the **Additional details** section below or can be found in the implementation of #1874.)
- Update the values of the Call To Action options in the `AssetGroupCard` because the actual values should be the small cases.
- Cast the type of the empty `assets` in the response of the asset groups fetching API from array to object. So that the `assets` will always be the JSON object type.

### Detailed test instructions:

1. Go to the campaign creation page.
2. Check if the "Continue" is enabled after entering valid audiences and the budget.
3. Go to the second step.
4. Check if the "Skip adding assets" can proceed with the campaign creation.

### Additional details:

An example data structure of `GET assets/suggestions?id=ID` response:

```json
{
  "headline": [
    "Beanie with Logo"
  ],
  "long_headline": [
    "Cat Pota: Beanie with Logo"
  ],
  "description": [
    "This is a simple product.",
    "Just another WordPress site"
  ],
  "logo": [],
  "business_name": "Cat Pota",
  "square_marketing_image": [
    "https://wp.test/wp-content/uploads/2021/01/beanie-with-logo-1.jpg"
  ],
  "marketing_image": [
    "https://wp.test/wp-content/uploads/2021/01/beanie-with-logo-1-800x418.jpg"
  ],
  "call_to_action_selection": null,
  "display_url_path": [
    "beanie-with-logo"
  ],
  "final_url": "https://wp.test/product/beanie-with-logo/"
}
```

An example data structure of `GET ads/campaigns/asset-groups?campaign_id=ID` response:

```json
[
  {
    "id": 6459681012,
    "final_url": "https://easonsu.a8c-localtunnel.cyou",
    "display_url_path": [
      "path-1",
      "path-2"
    ],
    "assets": {
      "square_marketing_image": [
        {
          "id": 63949253641,
          "content": "https://tpc.googlesyndication.com/simgad/13484709713110638229"
        }
      ],
      "marketing_image": [
        {
          "id": 63960463580,
          "content": "https://tpc.googlesyndication.com/simgad/6228362817145174537"
        }
      ],
      "logo": [
        {
          "id": 63960606224,
          "content": "https://tpc.googlesyndication.com/simgad/4585804890802605392"
        }
      ],
      "headline": [
        {
          "id": 67311903986,
          "content": "headline-1"
        },
        {
          "id": 67311903989,
          "content": "headline-2"
        },
        {
          "id": 67311903992,
          "content": "headline-3"
        }
      ],
      "long_headline": [
        {
          "id": 67311904001,
          "content": "long-headline"
        }
      ],
      "business_name": {
        "id": 67311904004,
        "content": "Cat Pota"
      },
      "description": [
        {
          "id": 70205253686,
          "content": "description-1"
        },
        {
          "id": 70205253689,
          "content": "description-2"
        }
      ]
    }
  }
]
```

### Changelog entry
